File: swift-bridge/Sources/SystemConfigurationBridge/ConsoleUser.swift
import Foundation
import SystemConfiguration

final class ConsoleUserBox {
    let name: String
    let uid: UInt32
    let gid: UInt32

    init(name: String, uid: uid_t, gid: gid_t) {
        self.name = name
        self.uid = UInt32(uid)
        self.gid = UInt32(gid)
    }
}

private func consoleUser(_ raw: UnsafeMutableRawPointer?) -> ConsoleUserBox? {
    unretained(raw)
}

@_cdecl("sc_console_user_copy_current")
public func sc_console_user_copy_current() -> UnsafeMutableRawPointer? {
    var uid: uid_t = 0
    var gid: gid_t = 0
    guard let name = SCDynamicStoreCopyConsoleUser(nil, &uid, &gid) as String? else {
        return nil
    }
    return retain(ConsoleUserBox(name: name, uid: uid, gid: gid))
}

@_cdecl("sc_console_user_copy_name")
public func sc_console_user_copy_name(_ raw: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let box = consoleUser(raw) else {
        return nil
    }
    return boxString(box.name)
}

@_cdecl("sc_console_user_get_uid")
public func sc_console_user_get_uid(_ raw: UnsafeMutableRawPointer?) -> UInt32 {
    guard let box = consoleUser(raw) else {
        return 0
    }
    return box.uid
}

@_cdecl("sc_console_user_get_gid")
public func sc_console_user_get_gid(_ raw: UnsafeMutableRawPointer?) -> UInt32 {
    guard let box = consoleUser(raw) else {
        return 0
    }
    return box.gid
}

File: src/console_user.rs
use crate::{bridge, ffi};

#[derive(Clone, Debug, Eq, PartialEq)]
/// Wraps `SCDynamicStoreCopyConsoleUser` results.
pub struct ConsoleUser {
    name: String,
    uid: u32,
    gid: u32,
}

impl ConsoleUser {
    /// Wraps `SCConsoleUserCopyCurrent`.
    pub fn current() -> Option<Self> {
        let raw = unsafe { ffi::console_user::sc_console_user_copy_current() };
        let raw = unsafe { bridge::OwnedHandle::from_raw(raw) }?;
        let name = bridge::take_optional_string(unsafe {
            ffi::console_user::sc_console_user_copy_name(raw.as_ptr())
        })?;
        let uid = unsafe { ffi::console_user::sc_console_user_get_uid(raw.as_ptr()) };
        let gid = unsafe { ffi::console_user::sc_console_user_get_gid(raw.as_ptr()) };
        Some(Self { name, uid, gid })
    }

    /// Wraps a helper on `SCDynamicStoreCopyConsoleUser`.
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Wraps a helper on `SCDynamicStoreCopyConsoleUser`.
    pub fn uid(&self) -> u32 {
        self.uid
    }

    /// Wraps a helper on `SCDynamicStoreCopyConsoleUser`.
    pub fn gid(&self) -> u32 {
        self.gid
    }
}
